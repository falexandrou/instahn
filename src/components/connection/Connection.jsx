import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { networkOnline, networkOffline } from 'actions/connection';

@connect( state => ({
  isOnline: state.connection.isOnline || false,
}))
class Connection extends React.Component {

  static propTypes = {
    isOnline: PropTypes.bool.isRequired,
  };

  componentDidMount() {
    const { dispatch } = this.props;

    this.setOnlineStatus();

    window.addEventListener('online', this.setOnline.bind(this));
    window.addEventListener('offline', this.setOffline.bind(this));
  }

  componentWillUnmount() {
    window.removeEventListener('online', this.setOnline.bind(this));
    window.removeEventListener('offline', this.setOffline.bind(this));
  }

  setOnlineStatus() {
    const isOnline = navigator.onLine;

    if (isOnline)
      return this.setOnline();

    return this.setOffline();
  }

  setOnline() {
    const { dispatch } = this.props;
    dispatch(networkOnline());
  }

  setOffline() {
    const { dispatch } = this.props;
    dispatch(networkOffline());
  }

  renderIndication() {
    const { isOnline } = this.props;

    if ( ! isOnline )
      return <span className="tag is-warning">Offline :(</span>;

    return <span className="tag is-primary">Online!</span>;
  }

  render() {
    return <div className="tags has-addons">
      <span className="tag">Connectivity</span>
      { this.renderIndication() }
    </div>;
  }
}

export default Connection;