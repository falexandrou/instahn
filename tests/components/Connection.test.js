import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Connection from 'components/connection/Connection';
import { Provider } from 'react-redux';

import { createStore } from '../mocks/store';

const fakeStore = createStore();

describe('Connection Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Provider store={fakeStore}><Connection /></Provider>, div);
  });

  it('should match snapshot when online', () => {
    let component = <Provider store={fakeStore}><Connection /></Provider>;
    const output = renderer.create(component).toJSON();
    expect(output).toMatchSnapshot();
  });

  it('should match snapshot when offline', () => {
    // Mock offline behavior
    Object.defineProperty(global.navigator, 'onLine', { value: false, writable: true });
    let component = <Provider store={fakeStore}><Connection /></Provider>;

    const output = renderer.create(component).toJSON();
    expect(output).toMatchSnapshot();
  });

});