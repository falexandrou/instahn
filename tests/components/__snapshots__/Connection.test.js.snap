// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection Component should match snapshot when offline 1`] = `
<div
  className="tags has-addons"
>
  <span
    className="tag"
  >
    Connectivity
  </span>
  <span
    className="tag is-warning"
  >
    Offline :(
  </span>
</div>
`;

exports[`Connection Component should match snapshot when online 1`] = `
<div
  className="tags has-addons"
>
  <span
    className="tag"
  >
    Connectivity
  </span>
  <span
    className="tag is-primary"
  >
    Online!
  </span>
</div>
`;
